android/src/main/kotlin/de/lschmierer/android_play_install_referrer/AndroidPlayInstallReferrerPlugin.kt

```diff
@@ -4,100 +4,95 @@ import android.content.Context
 import com.android.installreferrer.api.InstallReferrerStateListener
-import com.android.installreferrer.api.ReferrerDetails
 import io.flutter.embedding.engine.plugins.FlutterPlugin
-import io.flutter.plugin.common.BinaryMessenger
 import io.flutter.plugin.common.MethodCall
 import io.flutter.plugin.common.MethodChannel
 import io.flutter.plugin.common.MethodChannel.MethodCallHandler
 import io.flutter.plugin.common.MethodChannel.Result
-import io.flutter.plugin.common.PluginRegistry.Registrar
+import kotlinx.coroutines.*
 
 
+@ExperimentalCoroutinesApi
 /** AndroidPlayInstallReferrerPlugin */
 class AndroidPlayInstallReferrerPlugin : FlutterPlugin, MethodCallHandler {
-    private val CHANNEL_NAME = "de.lschmierer.android_play_install_referrer"
+
+    companion object {
+        const val CHANNEL_NAME = "de.lschmierer.android_play_install_referrer"
+    }
 
     private lateinit var context: Context
     private var channel: MethodChannel? = null
 
     override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
-        setupChannel(flutterPluginBinding.binaryMessenger, flutterPluginBinding.applicationContext)
-    }
-
-    // This static function is optional and equivalent to onAttachedToEngine. It supports the old
-    // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
-    // plugin registration via this function while apps migrate to use the new Android APIs
-    // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
-    //
-    // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
-    // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
-    // depending on the user's project. onAttachedToEngine or registerWith must both be defined
-    // in the same class.
-    companion object {
-        @JvmStatic
-        fun registerWith(registrar: Registrar) {
-            AndroidPlayInstallReferrerPlugin().setupChannel(registrar.messenger(), registrar.context())
-        }
+        this.context = flutterPluginBinding.applicationContext
+        channel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL_NAME)
+        channel!!.setMethodCallHandler(this)
     }
 
-    private fun setupChannel(messenger: BinaryMessenger, context: Context) {
-        this.context = context
-        channel = MethodChannel(messenger, CHANNEL_NAME)
-        channel!!.setMethodCallHandler(this)
+    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
+        channel?.setMethodCallHandler(null)
+        channel = null
     }
 
     override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
         if (call.method == "getInstallReferrer") {
-            getInstallReferrer(call, result)
+            val client = InstallReferrerClient.newBuilder(context).build()
+            runBlocking {
```

Comment:
Not sure if its a good ide to block the platform thread
```
Even though blocking the platform thread for inordinate amounts of time will not block the Flutter rendering pipeline, platforms do impose restrictions on expensive operations on this thread. So it is advised that any expensive work in response to platform messages be performed on separate worker threads (unrelated to the four threads discussed above) before having the responses queued back on the the platform thread for submission to the engine. Not doing so may result in platform-specific watchdogs terminating the application. Embeddings such as Android and iOS also uses the platform thread to pipe through user input events. A blocked platform thread can also cause gestures to be dropped.
```
https://github.com/flutter/flutter/wiki/The-Engine-architecture#platform-task-runner

Reply:
There is no changes from threading point of view. You can check implementation of InstallReferrerClientImpl and make sure that there is no thread switching inside. So both implementations are synchronised.

Reply:
Sure, this implementation is thread-safe. But now `onMethodCall` blocks for an undefined amount of time during which no other Flutter <-> Platform communication can take place.

I.e. `Embeddings such as Android and iOS also uses the platform thread to pipe through user input events. A blocked platform thread can also cause gestures to be dropped.`
https://github.com/flutter/flutter/wiki/The-Engine-architecture#platform-task-runner

Reply:
There was no intention from my side to change this behaviour. So initial implementation blocks thread (because all calls are synchronised) and final implementation also blocks thread. I don't see here heavy operation that would block thread for dramatic amount of time.

Reply:
The implementation currently on master does not block. `InstallReferrerStateListener ` is called asynchronously.

`runBlocking` runs a new coroutine and blocks the current thread until its completion.

Reply:
Sorry I can't agree with that.  
I am following sources from "prepare 0.1.1 release" commit.
`MethodCallHandler.onMethodCall` runs in UI thread.  
Plugin setup `referrerClient` variable there and runs `referrerClient.startConnection` with listener as argument.
There is no asynchronous code even if we will go to `InstallReferrerClientImpl` implementation which calls listener directly in `InstallReferrerClientImpl.startConnection` without any thread switching. 
Could you please check source code and explain me what do you mean by  `called asynchronously`?

Reply:
We pass a callback to `startConnection` which is called by the Install Referrer API later at some time (thus asynchronously). 
What your implementation essentially is doing is _waiting on the current thread for the callback to be called_ (thus blocking).

Reply:
> We pass a callback to startConnection which is called by the Install Referrer API later at some point in time (thus asynchronously). onMethodCall then returns before the callback is invoked.

Not true. In general callback could be called as synchronous so as asynchronous. In current implementation of `com.android.installreferrer:installreferrer:2.2` it is called synchronously. So implementation, that you already have, blocks thread. 

> What your implementation essentially is doing is waiting on the current thread for the callback to be called (thus blocking). The platform thread can not doing anything else during that time.

That's true. But in that point there is no difference between old and new implementation. As I already wrote, there was no intention to change that point.

Reply:
I consider the in inner workings of the Install Referrer API an implementation detail which we should not rely on.
The API documentation literally says

> Starts up InstallReferrerClient setup process asynchronously.

https://developer.android.com/reference/com/android/installreferrer/api/InstallReferrerClient

Btw. how do you access installreferrer source code?

> That's true. But in that point there is no difference between old and new implementation. As I already wrote, there was no intention to change that point.

So if there is no difference, what is the point of your implementation anyway? The (experimental) Kolin coroutine stuff makes the implementation just more verbose. I don't see the benefit.

Maybe you can reply to https://github.com/lschmierer/android_play_install_referrer/pull/16#issuecomment-994942518 and explain what the cause for the DeadObjectException is in the first place? So we can figure out a solution together.

Reply:
> Btw. how do you access installreferrer source code?

Decompiled classes in Android Studio.

> So if there is no difference, what is the point of your implementation anyway? The (experimental) Kolin coroutine stuff makes the implementation just more verbose. I don't see the benefit.

I have implemented first solution with saving references to client and result. You pointed me that it is not thread safe solution so I changed it in a way to have thread safety. Coroutines allow you to have thread safe solution. I agree that we can change it from `async` to `launch` and do not block main thread. So it is verbose but correct.  

Your new solution with HashMap is not thread safe because HashMap is not synchronised and can't be shared between threads without proper synchronisation.